BasicHistoryApi/app/src/main/java/com/google/android/gms/fit/samples/basichistoryapi/MainActivity.kt

```diff
@@ -0,0 +1,431 @@
+import com.google.android.gms.fit.samples.common.logger.MessageOnlyLogFilter
+import com.google.android.gms.fitness.Fitness
+import com.google.android.gms.fitness.FitnessOptions
+import com.google.android.gms.fitness.data.DataPoint
+import com.google.android.gms.fitness.data.DataSet
+import com.google.android.gms.fitness.data.DataSource
+import com.google.android.gms.fitness.data.DataType
+import com.google.android.gms.fitness.data.Field
+import com.google.android.gms.fitness.request.DataDeleteRequest
+import com.google.android.gms.fitness.request.DataReadRequest
+import com.google.android.gms.fitness.request.DataUpdateRequest
+import com.google.android.gms.fitness.result.DataReadResponse
+import com.google.android.gms.tasks.Task
+import java.text.DateFormat
+import java.util.Calendar
+import java.util.Date
+import java.util.TimeZone
+import java.util.concurrent.TimeUnit
+
+const val TAG = "BasicHistoryApi"
+
+/**
+ * This sample demonstrates how to use the History API of the Google Fit platform to insert data,
+ * query against existing data, and remove data. It also demonstrates how to authenticate a user
+ * with Google Play Services and how to properly represent data in a {@link DataSet}.
+ */
+class MainActivity : AppCompatActivity() {
+    private val dateFormat = DateFormat.getDateInstance()
+    private val fitnessOptions: FitnessOptions by lazy {
+        FitnessOptions.builder()
+                .addDataType(DataType.TYPE_STEP_COUNT_DELTA, FitnessOptions.ACCESS_WRITE)
+                .addDataType(DataType.AGGREGATE_STEP_COUNT_DELTA, FitnessOptions.ACCESS_WRITE)
+                .build()
+    }
+
+    // Two maps are used in conjunction with the OAuth flow. This is in order to identify what
+    // function to execute once the flow has successfully completed.
+    private val signInActionMap = mapOf<() -> Any, Int>(
+            ::insertAndReadData to 1,
+            ::updateAndReadData to 2,
+            ::deleteData to 3
+    )
+    private val signInResponseMap = signInActionMap.entries.associate { (k, v) -> v to k }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        initializeLogging()
+
+        fitSignIn(::insertAndReadData)
+    }
+
+    /**
+     * Checks that the user is signed in, and if so, executes the specified function. If the user is
+     * not signed in, initiates the sign in flow, specifying the post-sign in function to execute.
+     *
+     * @param signInAction The function to execute if already signed in, or once signed in.
+     */
+    private fun fitSignIn(signInAction: () -> Any) {
+        if (!GoogleSignIn.hasPermissions(getGoogleAccount(), fitnessOptions)) {
+            val requestCode = signInActionMap[signInAction]
+            requestCode?.let {
+                GoogleSignIn.requestPermissions(
+                        this,
+                        requestCode,
```

Comment:
Should this be 'it'? Kind of a nit, you might have left it that way to make it more clear maybe?

Reply:
No, good point - it's just me still learning Kotlin :-)